grid2op/Opponent/OpponentWithConverter.py

```diff
@@ -0,0 +1,183 @@
+
+        Returns
+        -------
+        res: ``object``
+            Anything that will be used by the BaseOpponent to take decisions.
+
+        """
+        return self.action_space.convert_obs(observation)
+
+    def convert_act(self, encoded_attack):
+        """
+        This function will convert an "ecnoded action" that be of any types, to a valid action that can be ingested
+        by the environment.
+
+        Parameters
+        ----------
+        encoded_attack: ``object``
+            Anything that represents an action.
+
+        Returns
+        -------
+        attack: :grid2op.BaseAction.BaseAction`
+            A valid actions, represented as a class, that corresponds to the encoded action given as input.
+
+        """
+        return self.action_space.convert_act(encoded_attack)
+
+    def attack(self, observation, *args):
+        """
+        Standard method of an :class:`BaseOpponent`. There is no need to overload this function.
+
+        Parameters
+        ----------
+        observation: :class:`grid2op.Observation.Observation`
+            The current observation of the :class:`grid2op.Environment.Environment`
+
+        reward: ``float``
+            The current reward. This is the reward obtained by the previous action
+
+        done: ``bool``
+            Whether the episode has ended or not. Used to maintain gym compatibility
+
+        Returns
+        -------
+        res: :class:`grid2op.Action.Action`
+            The action chosen by the bot / controler / opponent.
+
+        """
+        transformed_observation = self.convert_obs(observation)
+        # Todo : convert the other arguments?
+        encoded_attack = self.my_attack(transformed_observation, *args)
+        return self.convert_act(encoded_attack)
+
+    @abstractmethod
+    def my_attack(self, transformed_observation, *args):
+        """
+        This method should be overide if this class is used. It is an "abstract" method.
+
+        If someone wants to make a opponent that handles different kinds of actions an observation.
+
+        Parameters
+        ----------
+        transformed_observation: ``object``
+            Anything that will be used to create an action. This is the results to the call of
+            :func:`OpponentWithConverter.convert_obs`. This is likely a numpy array.
+
+        reward: ``float``
+            The current reward. This is the reward obtained by the previous action
+
+        done: ``bool``
+            Whether the episode has ended or not. Used to maintain gym compatibility
+
+        Returns
+        -------
+        res: ``object``
+            A representation of an action in any possible format. This action will then be ingested and formatted into
+            a valid action with the :func:`OpponentWithConverter.convert_act` method.
+
+        """
+        transformed_action = None
+        return transformed_action
```

Comment:
too much duplicated code. and not tested

Reply:
done